src/west/commands/project.py

```diff
@@ -604,6 +671,69 @@ def _checkout(project, branch):
     _git(project, 'checkout ' + branch)
 
 
+def _special_project(name):
+    # Returns a Project instance for one of the special repositories in west/,
+    # so that we can reuse the project-related functions for them
+
+    return Project(
+        name,
+        'dummy URL for {} repository'.format(name),
+        'master',
+        os.path.join('west', name.lower()),  # Path
+        os.path.join(util.west_dir(), name.lower()),  # Absolute path
+        None  # Clone depth
+    )
+
+
+_FAILED_SELFUPDATE_MSG = """
+Failed to self-update (name-and-path), while running command '{}'. Please fix
+the state of the repository, or pass --no-selfupdate to 'west fetch/pull' to
+skip self-updating for the duration of the command)"""[1:]
+
+def _selfupdate(update_west, update_manifest):
+    # 'try' is a keyword
+    def attempt(project, cmd):
+        res = _git(project, cmd, capture_stdout=True, check=False)
+        if res.returncode:
+            # The Git command's stderr isn't redirected and will also be
+            # available
+            _die(project, _FAILED_SELFUPDATE_MSG.format(cmd))
+        return res.stdout
+
+    projects = []
+    if update_west:
+        projects.append(_special_project('West'))
+    if update_manifest:
+        projects.append(_special_project('manifest'))
+
+    for project in projects:
+        _dbg(project, 'Self-updating (name-and-path)',
+             level=log.VERBOSE_NORMAL)
+
+        # Fetch changes from upstream
+        attempt(project, 'fetch')
+
+        # Get the SHA of the last commit in common with the upstream branch
+        merge_base = attempt(project, 'merge-base HEAD remotes/origin/master')
+
+        # Get the current SHA of the upstream branch
+        head_sha = attempt(project, 'show-ref --hash remotes/origin/master')
```

Comment:
The logic here could be used for normal projects later as well, to check if they are up-to-date.